### DISCUSSION TITLE
Unexpected result when using random() expression with CASE

### DISCUSSION THREAD

**Discussion by hdammeye** [duckdb/duckdb#12546] [Q&A] [answered]
Hello,

while l'm searching how to get randomly one of four bit patterns, I get some strange result.
Maybe sombody can explain.

Using the most current version. but same behavior in 0.10.x , 
```sql
select version();
```

| version() |
|-----------|
| v1.0.0    |

Expression ```trunc(random() * 4 + 1))::INTEGER``` return integers 1..4:

```sql
select  distinct i
 ,count(*) over (partition by i) as N
 ,count(*) over () as total
from (  select (trunc(random() * 4 + 1))::INTEGER as i from generate_series(1,1000000) ) q
order by 1;
```

| i |   N    |  total  |
|--:|-------:|--------:|
| 1 | 250181 | 1000000 |
| 2 | 250634 | 1000000 |
| 3 | 249507 | 1000000 |
| 4 | 249678 | 1000000 |

When using the expression with CASE it returns more:

```sql
select count(*) AS "#FALSE" 
from (
select
case (trunc(random() * 4 + 1))::INTEGER
        when  1 then '0001'
        when  2 then '0010'
        when  3 then '0100'
        when  4 then '1000'
        else 'unknown'
      end as bits
from generate_series(1, 1000000) t(i)
)
where bits not in ('0001','0010','0100','1000')
;
```

| #FALSE |
|-------:|
| 316953 |

Can sombody explain this result?
(Same Select in PostgreSQL 16.2 returns 0 .)

Thanks a lot and best regards

Heiko

**Comment by soerenwolfers** [comment]
Each WHEN branch of the CASE expression re-evaluates the original expression. This gives the probabilities 25%, 19% = 25% * (100% - 25%, 14% = 25% * (100% - 19%), 10% = 25% * (100%- 14%) of landing in one of the WHEN branches and the probability 30% = 100% - 25% - 19% - 14% - 10% of landing in the ELSE branch.

I'll open an issue to be make sure whether this is intended. (I don't think it is.)

**Reply by soerenwolfers** [reply, marked as answer]
https://github.com/duckdb/duckdb/issues/12547

**Comment by hdammeye** [comment]
Thanks for opening an issue.
Re-evaluating the expression for each branch feels wrong.  Of course it delivers something different than
evaluating the expression once and comparing the result.

```sql
select  distinct i
 ,count(*) over (partition by i) as N
 ,count(*) over () as total 
from (
select
case 
        when  (trunc(random() * 2 + 1))::INTEGER = 1 then 1
        when  (trunc(random() * 2 + 1))::INTEGER = 2 then 2
      end as i
from generate_series(1, 1_000_000) t(i)
)
order by i;
```

| i |   N    |  total  |
|--:|-------:|--------:|
| 1 | 500486 | 1000000 |
| 2 | 249897 | 1000000 |
|   | 249617 | 1000000 |
